src/main/java/com/ecobike/dao/FileBikeDao.java

```
@@ -0,0 +1,80 @@
+ */
+public class FileBikeDao implements BikeDao {
+
+    private static final Communicator communicator = EcoBikeApplication.getCommunicator();
+    /**
+     * Path to data file.
+     */
+    private final Path file;
+
+    public FileBikeDao(Path file) {
+        this.file = file;
+    }
+
+    /**
+     * Method reads data from data source.
+     *
+     * @throws IllegalDataSourceException if no one Bike object parsed from file.
+     */
+    @Override
+    public List<Bike> readBikes() throws IllegalDataSourceException {
+        List<String> lines;
+        try {
+            lines = Files.readAllLines(file);
+        } catch (IOException e) {
+            throw new IllegalDataSourceException(e);
+        }
+        List<Bike> bikes = new ArrayList<>(lines.size());
+        List<String> wrongLinesInfo = new ArrayList<>();
+        for (int i = 0; i < lines.size(); i++) {
+            try {
+                String line = lines.get(i);
+                BikeParser parser = FileBikeParser.getBikeParser(line);
+                bikes.add(parser.parseBike(line));
+            } catch (RuntimeException e) {
+                wrongLinesInfo.add("Line No. " + (i + 1) + " has wrong format");
+            }
+        }
+        if (bikes.isEmpty()) {
+            throw new IllegalDataSourceException();
+        }
+        wrongLinesInfo.forEach(communicator::writeMessage);
+        communicator.writeMessage(bikes.size() + " bike items has been read from the file");
+        return bikes;
+    }
+
+    /**
+     * Method writes to file Bike objects in text format.
+     * Old data in file will be replaced with new one.
+     */
+    @Override
+    public void saveBikes(List<? extends Bike> bikes) {
+        List<String> dataToWrite = bikes.stream()
+                .map(Bike::toFileWriterString)
+                .collect(Collectors.toList());
+        try {
+            Files.write(file, dataToWrite);
+        } catch (IOException e) {
+            e.printStackTrace();
```

do not ignore the exception

---

fixed